### ISSUE TITLE
SelfClosingTag linter could 

### ISSUE THREAD

**Issue by aried3r** [Shopify/erb-lint#89]
https://github.com/Shopify/erb-lint/blob/9e356ad6f56bbe32338dbbe2bed6000528d3a17c/lib/erb_lint/linters/self_closing_tag.rb#L10-L11

For example the [`br`](https://www.w3.org/TR/html52/textlevel-semantics.html#the-br-element) and [`link`](https://www.w3.org/TR/html52/document-metadata.html#the-link-element) element are [void elements](https://www.w3.org/TR/html5/syntax.html#void-elements) which only have a start tag.

> Void elements only have a start tag; end tags must not be specified for void elements.

Implicitely it says that they are not self closing as it goes on for [foreign elements](https://www.w3.org/TR/html5/syntax.html#foreign-elements):

> Foreign elements must either have a start tag and an end tag, or a start tag that is marked as self-closing, in which case they must not have an end tag.

As for other elements such as the [`command`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/command) one, these maybe should go into their own linter, suggesting not to use obsolete elements?

This is of course assuming most people use HTML5 and not XHTML(5). WDYT?

**Comment by EiNSTeiN-** [comment]
I guess it depends which spec this gem should follow, whether it should be configurable, etc.

**Comment by aried3r** [comment]
Well, lets discuss it! :)

I have no hard numbers on this, so arguing is kinda hard, but from my personal experience I see a lot of HTML5 and not so much XHTML anymore. I only saw [some suggesting](https://softwareengineering.stackexchange.com/a/149843/285799) that HTML5 "won". Seeing as Shopify probably uses this linter, do you use XHTML or HTML5?

I vote for HTML5, personally. What do you think?

**Comment by kernow** [comment]
I think this rule should be configurable with the options of:

- Enforced, must use self closing
- Enforced, must not use self closing
- Disabled

We can then support HTML5 and XHTML with an enforced style and let people choose the style they want. At the moment we have had to disable this rule but would like to enforce not using self closing tags.

**Comment by aried3r** [comment]
Good idea, should it only contain the list of void elements listed in the HTML5 standard? I'd really like to not include deprecated elements such as `command`, [`xmp`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/xmp) etc.

> At the moment we have had to disable this rule but would like to enforce not using self closing tags.

Same here.

**Comment by kernow** [comment]
I guess it should go with current best practice, whatever that is. There could be another linter for detecting deprecated elements. Maybe it should include deprecated elements in this linter as this rule wouldn't have an opinion on what is deprecated or not which would be the responsibility of another linter?

**Comment by EiNSTeiN-** [comment]
> Shopify probably uses this linter, do you use XHTML or HTML5?

We have `<!DOCTYPE html>` in the admin area (and most other places?) so that would make it HTML5.

> I think this rule should be configurable with the options of:

I'm also in favor of making this configurable 👍 

> I'd really like to not include deprecated elements

Agreed with kernow, let's make a separate linter for that :)

**Comment by bzang** [comment]
We just implemented `erb-lint` and like @kernow, had to disable `SelfClosingTag`.

[`htmllint`](https://github.com/htmllint/htmllint/wiki/Options#tag-self-close) seems to have reasonable rules that lines up with what's being described here:

> Default: false
> 
> "always": Void elements must be self-closed with / (html4 style).
> "never": Void elements must not be self-closed with / (html5 style).
> false: No restriction.
> The void elements are area, base, br, col, embed, hr, img, input, keygen, link, menuitem, meta, param, source, track, and wbr.

**Comment by bzang** [comment]
Please check out the PR #117, I put together and let me know what your thoughts are. I tried to keep the styles and norms from the other linters.

**Comment by muan** [comment]
Looks like #117 has been merged. Any chance we can get a new release? We'd love to use this rule setting as well. :)

**Comment by atesgoral** [comment]
I wish the default would be "never", since we're most likely writing HTML5 most of the time?

**Comment by paracycle** [comment]
> I wish the default would be "never", since we're most likely writing HTML5 most of the time?

@atesgoral I think the default will be `never`: https://github.com/Shopify/erb-lint/pull/117/files#diff-586d9c7f941bd4ddb157fa5321d71ee6R10

**Comment by gravitystorm** [comment]
It would be great to see a new release! We've just started using erblint and run into both this and #119, both of which were already fixed and are just waiting on a new erblint release.
